File: anytalk-daemon/src/asr.rs
use crate::audio::AudioMsg;
use crate::config::AsrConfig;
use crate::ipc::{serialize_msg, ServerMsg};
use crate::protocol::{build_audio_only_request, build_full_client_request, parse_server_message};
use futures_util::{SinkExt, StreamExt};
use std::sync::Arc;
use std::time::Duration;
use tokio::sync::{mpsc, Mutex as TokioMutex, Notify};
use tokio::time::sleep;
use tokio_tungstenite::tungstenite::client::IntoClientRequest;
use tokio_tungstenite::tungstenite::protocol::Message;
use tokio_tungstenite::{MaybeTlsStream, WebSocketStream};
use tracing::{debug, error, info, warn};

pub type WsStream = WebSocketStream<MaybeTlsStream<tokio::net::TcpStream>>;

/// Manages a single "hot spare" connection.
pub struct ConnectionPool {
    /// The pre-connected stream.
    pub spare: Arc<TokioMutex<Option<WsStream>>>,
    /// Notify when the spare is consumed, so the background task can reconnect.
    notify_consumed: Arc<Notify>,
    /// Config to use for connecting.
    config: AsrConfig,
}

impl ConnectionPool {
    pub fn new(config: AsrConfig) -> Self {
        Self {
            spare: Arc::new(TokioMutex::new(None)),
            notify_consumed: Arc::new(Notify::new()),
            config,
        }
    }

    /// Takes the spare connection if available.
    pub async fn take(&self) -> Option<WsStream> {
        let mut lock = self.spare.lock().await;
        let stream = lock.take();
        if stream.is_some() {
            self.notify_consumed.notify_one();
        }
        stream
    }

    /// Background task to maintain the spare connection.
    pub async fn run_maintainer(self: Arc<Self>) {
        loop {
            // Check if we need a connection
            let needs_conn = {
                let lock = self.spare.lock().await;
                lock.is_none()
            };

            if needs_conn {
                info!("Pre-connecting to Doubao...");
                match connect_to_asr(&self.config).await {
                    Ok(stream) => {
                        info!("Pre-connection established. Ready.");
                        let mut lock = self.spare.lock().await;
                        *lock = Some(stream);
                    }
                    Err(e) => {
                        error!("Pre-connection failed: {}. Retrying in 3s...", e);
                        sleep(Duration::from_secs(3)).await;
                        continue;
                    }
                }
            }

            // Wait until consumed
            self.notify_consumed.notified().await;
            // Slight delay to avoid hammering if something is spiraling,
            // but short enough to be ready for the next phrase.
            sleep(Duration::from_millis(100)).await;
        }
    }
}

pub async fn connect_to_asr(cfg: &AsrConfig) -> Result<WsStream, String> {
    let url = asr_url(&cfg.mode);
    debug!("Dialing ASR: {}", url);
    let mut request = url
        .into_client_request()
        .map_err(|e| format!("ws request error: {e}"))?;
    {
        let headers = request.headers_mut();
        headers.insert(
            "X-Api-App-Key",
            cfg.app_id.parse().map_err(|_| "bad app id")?,
        );
        headers.insert(
            "X-Api-Access-Key",
            cfg.access_token.parse().map_err(|_| "bad access token")?,
        );
        headers.insert(
            "X-Api-Resource-Id",
            cfg.resource_id.parse().map_err(|_| "bad resource id")?,
        );
        headers.insert(
            "X-Api-Connect-Id",
            uuid::Uuid::new_v4()
                .to_string()
                .parse()
                .map_err(|_| "bad uuid")?,
        );
    }

    let (ws_stream, _) = tokio_tungstenite::connect_async(request)
        .await
        .map_err(|e| format!("ws connect error: {e}"))?;
    Ok(ws_stream)
}

fn asr_url(mode: &str) -> &'static str {
    match mode {
        "bidi" => "wss://openspeech.bytedance.com/api/v3/sauc/bigmodel",
        "bidi_async" => "wss://openspeech.bytedance.com/api/v3/sauc/bigmodel_async",
        _ => "wss://openspeech.bytedance.com/api/v3/sauc/bigmodel_nostream",
    }
}

fn default_request_json(mode: &str) -> String {
    let is_nostream = mode == "nostream";
    let mut obj = serde_json::json!({
        "user": {"uid": "anytalk"},
        "audio": {
            "format": "pcm",
            "rate": 16000,
            "bits": 16,
            "channel": 1
        },
        "request": {
            "model_name": "bigmodel",
            "enable_itn": true,
            "enable_punc": true,
            "enable_ddc": false,
            "enable_word": false,
            "res_type": "full",
            "nbest": 1,
            "use_vad": true
        }
    });
    if is_nostream {
        if let Some(audio) = obj.get_mut("audio") {
            audio["language"] = serde_json::Value::String("zh-CN".to_string());
        }
    }
    obj.to_string()
}

pub async fn run_session(
    ws_stream: WsStream,
    mut audio_rx: mpsc::Receiver<AudioMsg>,
    resp_tx: mpsc::Sender<String>,
    cfg: AsrConfig,
) -> Result<(), String> {
    info!("Starting session on existing WS connection");
    let (mut ws_write, mut ws_read) = ws_stream.split();

    let req_json = default_request_json(&cfg.mode);
    debug!("Sending initial request: {}", req_json);
    let frame = build_full_client_request(&req_json);
    ws_write
        .send(Message::Binary(frame))
        .await
        .map_err(|e| format!("ws send error: {e}"))?;

    let mut last_committed_end_time: i64 = -1;
    let mut last_full_text = String::new();
    let mut chunk_count = 0;
    let mut audio_active = true;

    loop {
        tokio::select! {
            audio = audio_rx.recv(), if audio_active => {
                match audio {
                    Some(AudioMsg::Chunk(bytes)) => {
                        chunk_count += 1;
                        if chunk_count % 20 == 0 {
                            debug!("Sent 20 audio chunks to ASR...");
                        }
                        let frame = build_audio_only_request(&bytes, false);
                        if ws_write.send(Message::Binary(frame)).await.is_err() {
                            audio_active = false;
                        }
                    }
                    None => {
                        debug!("Audio source channel closed (Stop received)");
                        // IMPORTANT: Send an empty chunk with last=true to tell ASR we are done.
                        let empty = Vec::new();
                        let frame = build_audio_only_request(&empty, true);
                        if let Err(e) = ws_write.send(Message::Binary(frame)).await {
                             warn!("Failed to send final frame: {}", e);
                        }
                        audio_active = false;
                    }
                }
            }
            msg = ws_read.next() => {
                match msg {
                    Some(Ok(Message::Binary(data))) => {
                        let parsed = parse_server_message(&data);
                        if parsed.kind == "error" {
                            let msg = parsed.error_msg.unwrap_or_else(|| "server error".to_string());
                            error!("ASR Error: {}", msg);
                            let _ = resp_tx.send(serialize_msg(ServerMsg::Error { message: &msg })).await;
                            break;
                        }
                        if parsed.kind != "response" {
                            continue;
                        }
                        if let Some(json_text) = parsed.json_text {
                            debug!("ASR Response (flags={:b}): {}", parsed.flags, json_text);
                            let (partial, finals) = parse_asr_texts(&json_text, &mut last_committed_end_time, &mut last_full_text, cfg.mode.as_str());
                            if let Some(p) = partial {
                                let _ = resp_tx.send(serialize_msg(ServerMsg::Partial { text: &p })).await;
                            }
                            for f in finals {
                                debug!("Committing final text: {}", f);
                                let _ = resp_tx.send(serialize_msg(ServerMsg::Final { text: &f })).await;
                            }
                            // 0b0011 means this is the final response frame from server
                            if parsed.flags == 0b0011 {
                                info!("Received final server response frame. Closing.");
                                break;
                            }
                        }
                    }
                    Some(Ok(Message::Close(_))) => {
                        info!("WebSocket closed by server");
                        break;
                    }
                    Some(Ok(_)) => {},
                    Some(Err(e)) => {
                        error!("WebSocket error: {}", e);
                        break;
                    }
                    None => {
                        debug!("WebSocket stream ended (None)");
                        break;
                    }
                }
            }
        }
    }

    Ok(())
}

fn parse_asr_texts(
    json_text: &str,
    last_committed_end_time: &mut i64,
    last_full_text: &mut String,
    mode: &str,
) -> (Option<String>, Vec<String>) {
    let mut partial: Option<String> = None;
    let mut finals: Vec<String> = Vec::new();

    let obj: serde_json::Value = match serde_json::from_str(json_text) {
        Ok(v) => v,
        Err(_) => return (None, finals),
    };
    let result = match obj.get("result") {
        Some(r) => r,
        None => return (None, finals),
    };

    if let Some(utterances) = result.get("utterances").and_then(|u| u.as_array()) {
        for u in utterances {
            let def = u.get("definite").and_then(|v| v.as_bool()).unwrap_or(false);
            if !def {
                continue;
            }
            let end_time = u.get("end_time").and_then(|v| v.as_i64()).unwrap_or(-1);
            if end_time <= *last_committed_end_time {
                debug!(
                    "Skipping definite utterance: end_time {} <= last {}",
                    end_time,
                    last_committed_end_time
                );
                continue;
            }
            if let Some(txt) = u.get("text").and_then(|v| v.as_str()) {
                let trimmed = txt.trim();
                if !trimmed.is_empty() {
                    debug!("New final: {} (end_time {})", trimmed, end_time);
                    finals.push(trimmed.to_string());
                    *last_committed_end_time = end_time;
                }
            }
        }
        for u in utterances.iter().rev() {
            if u.get("definite").and_then(|v| v.as_bool()).unwrap_or(false) {
                continue;
            }
            if let Some(txt) = u.get("text").and_then(|v| v.as_str()) {
                let trimmed = txt.trim();
                if !trimmed.is_empty() {
                    partial = Some(trimmed.to_string());
                    break;
                }
            }
        }
        return (partial, finals);
    }

    if let Some(txt) = result.get("text").and_then(|v| v.as_str()) {
        let full = txt.trim().to_string();
        if full.is_empty() {
            return (None, finals);
        }
        if mode == "bidi_async" {
            partial = Some(full.clone());
            finals.push(full.clone());
        } else if !last_full_text.is_empty() && full.starts_with(last_full_text.as_str()) {
            let suffix = full[last_full_text.len()..].trim();
            if !suffix.is_empty() {
                finals.push(suffix.to_string());
            }
        } else if full != *last_full_text {
            finals.push(full.clone());
        }
        *last_full_text = full;
    }

    (partial, finals)
}

File: anytalk-lib/src/asr.zig
const std = @import("std");
const WebSocket = @import("websocket.zig").WebSocket;
const protocol = @import("protocol.zig");
const json_mod = @import("json.zig");
const uuid_mod = @import("uuid.zig");
const audio_mod = @import("audio.zig");
const log = std.log.scoped(.asr);

pub const AsrConfig = struct {
    app_id: []const u8,
    access_token: []const u8,
    resource_id: []const u8,
    mode: []const u8,
};

fn asrUrl(mode: []const u8) struct { host: []const u8, path: []const u8 } {
    if (std.mem.eql(u8, mode, "bidi")) {
        return .{ .host = "openspeech.bytedance.com", .path = "/api/v3/sauc/bigmodel" };
    } else if (std.mem.eql(u8, mode, "bidi_async")) {
        return .{ .host = "openspeech.bytedance.com", .path = "/api/v3/sauc/bigmodel_async" };
    } else {
        return .{ .host = "openspeech.bytedance.com", .path = "/api/v3/sauc/bigmodel_nostream" };
    }
}

pub fn connectToAsr(allocator: std.mem.Allocator, cfg: *const AsrConfig) !WebSocket {
    const url = asrUrl(cfg.mode);
    const connect_id = uuid_mod.v4();

    const headers = [_][2][]const u8{
        .{ "X-Api-App-Key", cfg.app_id },
        .{ "X-Api-Access-Key", cfg.access_token },
        .{ "X-Api-Resource-Id", cfg.resource_id },
        .{ "X-Api-Connect-Id", &connect_id },
    };

    log.info("Connecting to ASR: {s}{s}", .{ url.host, url.path });
    return WebSocket.connect(allocator, url.host, 443, url.path, &headers);
}

/// Hot spare connection pool - maintains one pre-connected WebSocket.
pub const ConnectionPool = struct {
    spare: ?WebSocket,
    mutex: std.Thread.Mutex,
    allocator: std.mem.Allocator,
    cfg: AsrConfig,
    thread: ?std.Thread,
    running: std.atomic.Value(bool),
    consumed: std.Thread.ResetEvent,

    pub fn init(allocator: std.mem.Allocator, cfg: AsrConfig) ConnectionPool {
        return ConnectionPool{
            .spare = null,
            .mutex = .{},
            .allocator = allocator,
            .cfg = cfg,
            .thread = null,
            .running = std.atomic.Value(bool).init(false),
            .consumed = .{},
        };
    }

    pub fn start(self: *ConnectionPool) !void {
        if (self.running.load(.acquire)) return;
        self.running.store(true, .release);
        self.thread = try std.Thread.spawn(.{}, maintainerLoop, .{self});
    }

    pub fn stop(self: *ConnectionPool) void {
        self.running.store(false, .release);
        self.consumed.set(); // Wake up maintainer so it can exit
        if (self.thread) |t| {
            t.join();
            self.thread = null;
        }
        self.mutex.lock();
        defer self.mutex.unlock();
        if (self.spare) |*ws| {
            ws.close();
            self.spare = null;
        }
    }

    pub fn take(self: *ConnectionPool) ?WebSocket {
        self.mutex.lock();
        defer self.mutex.unlock();
        const ws = self.spare;
        if (ws != null) {
            self.spare = null;
            self.consumed.set();
        }
        return ws;
    }

    fn maintainerLoop(self: *ConnectionPool) void {
        log.info("Connection pool maintainer started", .{});
        while (self.running.load(.acquire)) {
            const needs_conn = blk: {
                self.mutex.lock();
                defer self.mutex.unlock();
                break :blk self.spare == null;
            };

            if (needs_conn) {
                log.info("Pre-connecting to Doubao...", .{});
                if (connectToAsr(self.allocator, &self.cfg)) |ws| {
                    log.info("Pre-connection established. Ready.", .{});
                    self.mutex.lock();
                    defer self.mutex.unlock();
                    self.spare = ws;
                } else |err| {
                    log.err("Pre-connection failed: {any}. Retrying in 3s...", .{err});
                    std.Thread.sleep(3 * std.time.ns_per_s);
                    continue;
                }
            }

            // Wait until consumed
            self.consumed.reset();
            self.consumed.timedWait(30 * std.time.ns_per_s) catch {};
            std.Thread.sleep(100 * std.time.ns_per_ms);
        }
        log.info("Connection pool maintainer stopped", .{});
    }
};

/// Represents an active ASR session. Runs in its own thread.
pub const Session = struct {
    ws: WebSocket,
    allocator: std.mem.Allocator,
    cfg: AsrConfig,
    callback: *const fn (event_type: u32, text: [*c]const u8, user_data: ?*anyopaque) void,
    user_data: ?*anyopaque,
    audio_target: *audio_mod.AudioTarget,
    thread: ?std.Thread,
    running: std.atomic.Value(bool),
    /// Ring buffer for audio chunks from capture thread
    audio_ring: AudioRing,

    pub fn init(
        ws: WebSocket,
        allocator: std.mem.Allocator,
        cfg: AsrConfig,
        callback: *const fn (event_type: u32, text: [*c]const u8, user_data: ?*anyopaque) void,
        user_data: ?*anyopaque,
        audio_target: *audio_mod.AudioTarget,
    ) Session {
        return Session{
            .ws = ws,
            .allocator = allocator,
            .cfg = cfg,
            .callback = callback,
            .user_data = user_data,
            .audio_target = audio_target,
            .thread = null,
            .running = std.atomic.Value(bool).init(false),
            .audio_ring = .{},
        };
    }

    pub fn start(self: *Session) !void {
        self.running.store(true, .release);
        // Set audio target to route audio into our ring buffer
        self.audio_target.setTarget(audioCallback, @ptrCast(self));
        self.thread = try std.Thread.spawn(.{}, sessionLoop, .{self});
    }

    pub fn stopAudio(self: *Session) void {
        self.audio_target.clearTarget();
    }

    pub fn cancel(self: *Session) void {
        self.running.store(false, .release);
        self.audio_target.clearTarget();
    }

    pub fn join(self: *Session) void {
        if (self.thread) |t| {
            t.join();
            self.thread = null;
        }
    }

    fn audioCallback(data: []const u8, user_data: ?*anyopaque) void {
        const self: *Session = @ptrCast(@alignCast(user_data.?));
        self.audio_ring.push(data);
    }

    fn emitEvent(self: *Session, event_type: u32, text: []const u8) void {
        // Create null-terminated copy
        var buf: [4096]u8 = undefined;
        const len = @min(text.len, buf.len - 1);
        @memcpy(buf[0..len], text[0..len]);
        buf[len] = 0;
        self.callback(event_type, &buf, self.user_data);
    }

    fn sessionLoop(self: *Session) void {
        self.runSession() catch |err| {
            log.err("Session error: {any}", .{err});
            self.emitEvent(3, "session error"); // ANYTALK_EVENT_ERROR
        };
        // Session done - emit idle status
        self.emitEvent(2, "idle"); // ANYTALK_EVENT_STATUS
    }

    fn runSession(self: *Session) !void {
        log.info("Starting ASR session", .{});

        // Set a read timeout so the loop can check the running flag periodically
        self.ws.tls.setReadTimeout(200);

        // Send initial request JSON
        const req_json = try json_mod.buildInitialRequest(self.allocator, self.cfg.mode);
        defer self.allocator.free(req_json);

        const initial_frame = try protocol.buildFullClientRequest(self.allocator, req_json);
        defer self.allocator.free(initial_frame);

        self.ws.sendBinary(initial_frame) catch |err| {
            log.err("Failed to send initial request: {any}", .{err});
            return err;
        };

        var last_committed_end_time: i64 = -1;
        var last_full_text: []const u8 = "";
        var chunk_count: u32 = 0;
        var audio_done = false;

        while (self.running.load(.acquire)) {
            // Try to read audio and send it
            if (!audio_done) {
                if (self.audio_ring.pop()) |chunk| {
                    chunk_count += 1;
                    if (chunk_count % 20 == 0) {
                        log.info("Sent {d} audio chunks to ASR...", .{chunk_count});
                    }
                    const frame = protocol.buildAudioOnlyRequest(self.allocator, &chunk, false) catch continue;
                    defer self.allocator.free(frame);
                    self.ws.sendBinary(frame) catch {
                        audio_done = true;
                    };
                } else {
                    // No audio available, check if target was cleared (stop signal)
                    if (!self.audio_target.isActive()) {
                        log.info("Audio source stopped (Stop received)", .{});
                        // Send empty final frame
                        const final_frame = protocol.buildAudioOnlyRequest(self.allocator, &.{}, true) catch {
                            audio_done = true;
                            continue;
                        };
                        defer self.allocator.free(final_frame);
                        self.ws.sendBinary(final_frame) catch {};
                        audio_done = true;
                    }
                }
            }

            // Try to read WebSocket response (200ms SO_RCVTIMEO on socket)
            const frame = self.ws.readFrame() catch |err| {
                switch (err) {
                    error.WouldBlock => continue, // Timeout, check running flag
                    error.ConnectionClosed => {
                        log.info("WebSocket closed", .{});
                        break;
                    },
                    else => {
                        log.err("WebSocket read error: {any}", .{err});
                        break;
                    },
                }
            };
            defer self.allocator.free(frame.payload);

            switch (frame.opcode) {
                .binary => {
                    const parsed = protocol.parseServerMessage(frame.payload);
                    if (parsed.kind == .@"error") {
                        const msg = parsed.error_msg orelse "server error";
                        log.err("ASR Error: {s}", .{msg});
                        self.emitEvent(3, msg);
                        break;
                    }
                    if (parsed.kind != .response) continue;
                    if (parsed.json_text) |jt| {
                        const result = json_mod.parseAsrResponse(
                            self.allocator,
                            jt,
                            &last_committed_end_time,
                            &last_full_text,
                            self.cfg.mode,
                        ) catch continue;

                        if (result.partial) |p| {
                            self.emitEvent(0, p); // PARTIAL
                            self.allocator.free(p);
                        }
                        for (result.finals) |f| {
                            self.emitEvent(1, f); // FINAL
                            self.allocator.free(f);
                        }
                        self.allocator.free(result.finals);

                        // 0b0011 means final server response frame
                        if (parsed.flags == 0b0011) {
                            log.info("Received final server response frame. Closing.", .{});
                            break;
                        }
                    }
                },
                .close => {
                    log.info("WebSocket closed by server", .{});
                    break;
                },
                .ping => {
                    self.ws.sendPong(frame.payload) catch {};
                },
                else => {},
            }
        }

        if (last_full_text.len > 0) {
            self.allocator.free(last_full_text);
        }

        self.ws.close();
    }
};

/// Simple lock-free ring buffer for audio chunks.
/// Each slot holds one 6400-byte chunk.
const AudioRing = struct {
    const SLOTS = 32;
    slots: [SLOTS][audio_mod.CHUNK_BYTES]u8 = undefined,
    write_pos: std.atomic.Value(usize) = std.atomic.Value(usize).init(0),
    read_pos: std.atomic.Value(usize) = std.atomic.Value(usize).init(0),

    fn push(self: *AudioRing, data: []const u8) void {
        const wp = self.write_pos.load(.acquire);
        const rp = self.read_pos.load(.acquire);
        // Drop if full
        if ((wp + 1) % SLOTS == rp) return;
        const slot = wp % SLOTS;
        const len = @min(data.len, audio_mod.CHUNK_BYTES);
        @memcpy(self.slots[slot][0..len], data[0..len]);
        self.write_pos.store((wp + 1) % SLOTS, .release);
    }

    fn pop(self: *AudioRing) ?[audio_mod.CHUNK_BYTES]u8 {
        const rp = self.read_pos.load(.acquire);
        const wp = self.write_pos.load(.acquire);
        if (rp == wp) return null;
        const slot = rp % SLOTS;
        const data = self.slots[slot];
        self.read_pos.store((rp + 1) % SLOTS, .release);
        return data;
    }
};
